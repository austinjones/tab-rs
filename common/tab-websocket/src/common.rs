use crate::WebSocket;
use futures::SinkExt;
use log::error;
use serde::Serialize;
use tungstenite::{Error, Message};

pub async fn send_message<Msg: Serialize>(websocket: &mut WebSocket, message: Msg) {
    let encoded = bincode::serialize(&message);

    if let Err(e) = encoded {
        error!("failed to encode message: {}", e);
        return;
    }

    let sent = websocket.send(Message::Binary(encoded.unwrap())).await;
    if let Err(e) = sent {
        error!("failed to send message: {}", e);
    }
}

pub async fn send_close(websocket: &mut WebSocket) {
    if let Err(e) = websocket.send(Message::Close(None)).await {
        match e {
            Error::ConnectionClosed | Error::AlreadyClosed => {
                return;
            }
            _ => {
                error!("failed to send close frame: {}", e);
            }
        }
    }
}
