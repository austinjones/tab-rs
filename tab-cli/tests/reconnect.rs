use insta::assert_snapshot;

mod common;
use common::*;

/// Tests that a session can be established, disconnected from, and re-established
/// Covers connection, ctrl-W, disconnection, and scrollback
#[tokio::test]
async fn reconnect() -> anyhow::Result<()> {
    let mut session = TestSession::new()?;

    let result = session
        .command()
        .tab("reconnect/")
        .delay_ms(1000)
        .stdin("echo foo\n")
        .delay_ms(200) // delay a few moment, so we can confirm `echo foo` is echoed
        .stdin_bytes(&[23u8])
        .run()
        .await?;

    assert_eq!(Some(0), result.exit_status.code());
    assert_snapshot!("before", result.stdout);

    let result = session
        .command()
        .tab("reconnect/")
        .delay_ms(500)
        .stdin("exit\n")
        .run()
        .await?;
    assert_eq!(Some(0), result.exit_status.code());
    assert_snapshot!("after", result.stdout);

    Ok(())
}
