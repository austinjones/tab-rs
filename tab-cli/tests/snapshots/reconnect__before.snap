---
source: tab-cli/tests/reconnect.rs
expression: output_string
---
"echo foo\r\r\n\u{1b}[?1034hsh-3.2$ echo foo\r\r\nfoo\r\r\nsh-3.2$ "
